### ISSUE TITLE
Unable to load synchronized event, frame and depth data

### ISSUE THREAD

**Issue by gaudelbijay** [anusha-devulapally/ER-F2D#2]
# DataLoader Mismatch: Event Frame and Depth Data Misalignment

## Description

I prepared the MVSEC data using the data preparation script provided below and attempted to use the current dataloader, but it wasn't successful. After trying to load the data with the testing script, the saved visualization images do not accurately correspond to the RGB, event, and depth data. Any assistance would be greatly appreciated.

## Steps to Reproduce

1. **Prepare MVSEC Data**

   Use the following script to preprocess the MVSEC outdoor dataset:

   ```python
   #!/usr/bin/env python3
   
   import os
   import sys
   import cv2
   import h5py
   import numpy as np
   from tqdm import tqdm
   
   def events_to_voxel_grid(events, num_bins=5, width=346, height=260):
       """Convert events to a voxel grid representation."""
       if len(events) == 0:
           return np.zeros((num_bins, height, width), dtype=np.float32)
   
       xs = events[:, 0].astype(np.int32)
       ys = events[:, 1].astype(np.int32)
       ts = events[:, 2].astype(np.float64)
       ps = events[:, 3].astype(np.float32)
   
       t_min, t_max = ts.min(), ts.max()
       t_norm = (ts - t_min) / (t_max - t_min) if t_max > t_min else np.zeros_like(ts)
   
       t_bin = (t_norm * (num_bins - 1)).astype(np.int32)
       voxel_grid = np.zeros((num_bins, height, width), dtype=np.float32)
   
       valid_mask = (
           (xs >= 0) & (xs < width) &
           (ys >= 0) & (ys < height) &
           (t_bin >= 0) & (t_bin < num_bins)
       )
   
       xs, ys, t_bin, ps = xs[valid_mask], ys[valid_mask], t_bin[valid_mask], ps[valid_mask]
       np.add.at(voxel_grid, (t_bin, ys, xs), ps)
       return voxel_grid
   
   def create_color_event_image(voxel_grid):
       """Create a colored visualization of the event data."""
       event_frame = np.sum(voxel_grid, axis=0)
       pos_vals = np.where(event_frame > 0, event_frame, 0)
       neg_vals = np.where(event_frame < 0, -event_frame, 0)
   
       max_val = max(pos_vals.max(), neg_vals.max())
       pos_norm = (pos_vals / max_val * 255).astype(np.uint8) if max_val > 0 else pos_vals.astype(np.uint8)
       neg_norm = (neg_vals / max_val * 255).astype(np.uint8) if max_val > 0 else neg_vals.astype(np.uint8)
   
       return np.stack([neg_norm, np.zeros_like(pos_norm), pos_norm], axis=-1)
   
   def initialize_timestamp_file(file_path):
       """Initialize or clear a timestamp file."""
       os.makedirs(os.path.dirname(file_path), exist_ok=True)
       open(file_path, 'w').close()
   
   def save_timestamp(file_path, timestamp):
       """
       Save timestamp in the format expected by EventDataset:
       index timestamp
       """
       if timestamp is not None:
           try:
               with open(file_path, 'r') as f:
                   index = sum(1 for _ in f)
           except FileNotFoundError:
               index = 0
               
           with open(file_path, 'a') as f:
               f.write(f"{index} {timestamp}\n")
       else:
           with open(file_path, 'a') as f:
               f.write("NaN NaN\n")
   
   def save_data(output_dir, index, voxel_grid=None, image=None, depth=None, flow=None, 
                 event_ts=None, image_ts=None, depth_ts=None, flow_ts=None):
       """Save all data types with their timestamps."""
       
       if voxel_grid is not None:
           voxels_dir = os.path.join(output_dir, "events/voxels")
           frame_color_dir = os.path.join(output_dir, "events/frame_color")
           os.makedirs(voxels_dir, exist_ok=True)
           os.makedirs(frame_color_dir, exist_ok=True)
   
           np.save(os.path.join(voxels_dir, f"event_tensor_{index:010d}.npy"), voxel_grid)
           save_timestamp(os.path.join(voxels_dir, "timestamps.txt"), event_ts)
   
           color_image = create_color_event_image(voxel_grid)
           cv2.imwrite(os.path.join(frame_color_dir, f"event_color_{index:010d}.png"), color_image)
           save_timestamp(os.path.join(frame_color_dir, "timestamps.txt"), event_ts)
   
       if image is not None:
           rgb_dir = os.path.join(output_dir, "rgb/davis")
           os.makedirs(rgb_dir, exist_ok=True)
           cv2.imwrite(os.path.join(rgb_dir, f"frame_{index:010d}.png"), image)
           save_timestamp(os.path.join(rgb_dir, "timestamps.txt"), image_ts)
   
       if depth is not None:
           depth_data_dir = os.path.join(output_dir, "depth/data")
           depth_frames_dir = os.path.join(output_dir, "depth/frames")
           os.makedirs(depth_data_dir, exist_ok=True)
           os.makedirs(depth_frames_dir, exist_ok=True)
   
           np.save(os.path.join(depth_data_dir, f"depth_{index:010d}.npy"), depth)
           save_timestamp(os.path.join(depth_data_dir, "timestamps.txt"), depth_ts)
   
           if np.isfinite(depth).all():
               depth_vis = ((depth - depth.min()) / (depth.max() - depth.min() + 1e-8) * 255).astype(np.uint8)
               depth_vis = cv2.applyColorMap(depth_vis, cv2.COLORMAP_JET)
               cv2.imwrite(os.path.join(depth_frames_dir, f"depth_{index:010d}.png"), depth_vis)
               save_timestamp(os.path.join(depth_frames_dir, "timestamps.txt"), depth_ts)
   
       if flow is not None:
           flow_data_dir = os.path.join(output_dir, "flow/data")
           flow_frames_dir = os.path.join(output_dir, "flow/frames")
           os.makedirs(flow_data_dir, exist_ok=True)
           os.makedirs(flow_frames_dir, exist_ok=True)
   
           np.save(os.path.join(flow_data_dir, f"flow_{index:010d}.npy"), flow)
           save_timestamp(os.path.join(flow_data_dir, "timestamps.txt"), flow_ts)
   
           flow_mag = np.linalg.norm(flow, axis=0)
           flow_vis = (flow_mag - flow_mag.min()) / (flow_mag.max() + 1e-8) * 255
           flow_vis = cv2.applyColorMap(flow_vis.astype(np.uint8), cv2.COLORMAP_JET)
           cv2.imwrite(os.path.join(flow_frames_dir, f"flow_{index:010d}.png"), flow_vis)
           save_timestamp(os.path.join(flow_frames_dir, "timestamps.txt"), flow_ts)
   
   def initialize_output_directories(output_path):
       """Initialize all output directories and their timestamp files."""
       timestamp_files = [
           os.path.join(output_path, "events/voxels/timestamps.txt"),
           os.path.join(output_path, "events/frame_color/timestamps.txt"),
           os.path.join(output_path, "rgb/davis/timestamps.txt"),
           os.path.join(output_path, "depth/data/timestamps.txt"),
           os.path.join(output_path, "depth/frames/timestamps.txt"),
           os.path.join(output_path, "flow/data/timestamps.txt"),
           os.path.join(output_path, "flow/frames/timestamps.txt")
       ]
       
       for ts_file in timestamp_files:
           initialize_timestamp_file(ts_file)
   
   def process_mvsec_outdoor(data_hdf5_path, gt_hdf5_path, output_path):
       """Process MVSEC outdoor dataset and save in the correct format."""
       print(f"\nProcessing MVSEC outdoor data:")
       print(f"  Data HDF5: {data_hdf5_path}")
       print(f"  GT HDF5:   {gt_hdf5_path}")
       print(f"  Output:    {output_path}\n")
   
       # Initialize all output directories and timestamp files
       initialize_output_directories(output_path)
   
       try:
           data_file = h5py.File(data_hdf5_path, 'r')
           gt_file = h5py.File(gt_hdf5_path, 'r')
       except Exception as e:
           print(f"Error opening HDF5 files: {e}")
           return False
   
       davis_left_data = data_file['davis']['left']
       davis_left_gt = gt_file['davis']['left']
   
       events = davis_left_data['events'][:]
       images = davis_left_data['image_raw']
       image_event_inds = davis_left_data['image_raw_event_inds']
   
       depths = davis_left_gt.get('depth_image_rect')
       flows = davis_left_gt.get('flow_dist')
   
       print(f"Total events: {len(events):,}")
       print(f"Frame count: {len(images)}")
       print(f"Depth count: {len(depths) if depths is not None else 0}")
       print(f"Flow count:  {len(flows) if flows is not None else 0}")
   
       image_ts_data = davis_left_data['image_raw_ts'][:]
       depth_ts_data = davis_left_gt.get('depth_image_rect_ts', [])
       flow_ts_data = davis_left_gt.get('flow_dist_ts', [])
   
       print("\nProcessing frames...")
       for i in tqdm(range(len(images) - 1)):
           start_idx = image_event_inds[i]
           end_idx = image_event_inds[i + 1]
           frame_events = events[start_idx:end_idx]
   
           voxel_grid = events_to_voxel_grid(frame_events)
   
           save_data(
               output_dir=output_path,
               index=i,
               voxel_grid=voxel_grid,
               image=images[i],
               depth=depths[i] if depths is not None and i < len(depths) else None,
               flow=flows[i] if flows is not None and i < len(flows) else None,
               event_ts=frame_events[-1, 2] if len(frame_events) > 0 else None,
               image_ts=image_ts_data[i] if i < len(image_ts_data) else None,
               depth_ts=depth_ts_data[i] if depths is not None and i < len(depth_ts_data) else None,
               flow_ts=flow_ts_data[i] if flows is not None and i < len(flow_ts_data) else None
           )
   
       print("\nProcessing completed successfully!")
       return True
   
   if __name__ == "__main__":
       # Define default paths - modify these as needed
       data_hdf5 = "../data/outdoor_day1_data.hdf5"
       gt_hdf5 = "../data/outdoor_day1_gt.hdf5"
       output_dir = "../data/day1/train"
   
       # Create output directory if it doesn't exist
       os.makedirs(output_dir, exist_ok=True)
   
       # Process the data
       success = process_mvsec_outdoor(
           data_hdf5_path=data_hdf5,
           gt_hdf5_path=gt_hdf5,
           output_path=output_dir
       )
   
       # Exit with appropriate status code
       sys.exit(0 if success else 1)


And test code is: 

```python  
  #!/usr/bin/env python3
  import os
  import torch
  import matplotlib.pyplot as plt
  import numpy as np
  from dataloader import build_dataset
  from torch.utils.data import DataLoader
  
  def check_data_structure():
      base_path = "./data/day1/train"
      paths_to_check = [
          "depth/data",
          "events/voxels",
          "rgb/davis"
      ]
      
      print("Checking data structure...")
      for path in paths_to_check:
          full_path = os.path.join(base_path, path)
          if os.path.exists(full_path):
              print(f"✓ Found {full_path}")
              files = os.listdir(full_path)
              print(f"  Contains {len(files)} files")
              print(f"  First few files: {files[:5]}")
              if "timestamps.txt" in files:
                  print(f"  ✓ Found timestamps.txt")
              else:
                  print(f"  ✗ Missing timestamps.txt")
          else:
              print(f"✗ Missing {full_path}")
    
  def normalize_image(img):
      """Normalize image to [0,1] range"""
      img = img.astype(np.float32)
      if img.max() > img.min():
          img = (img - img.min()) / (img.max() - img.min())
      return img
  
  def test_data():
      """Test data loading from the dataset"""
      
      class Args:
          def __init__(self):
              self.data_path = "./data/day1/"
              self.train_step_size = 1
              self.val_step_size = 1
              self.clip_distance = 100.0
              self.reg_factor = 3.7
  
      args = Args()
  
      print("Building dataset...")
      dataset = build_dataset('train', None, args)
      print(f"Dataset size: {len(dataset)}")
  
      dataloader = DataLoader(dataset, batch_size=1, shuffle=False)
  
      print("\nTesting first batch...")
      try:
          batch_data = next(iter(dataloader))
          sample_dict = batch_data[0]
  
          # Print data statistics
          print("\nData ranges:")
          if isinstance(sample_dict, dict):
              for key, value in sample_dict.items():
                  if isinstance(value, torch.Tensor):
                      print(f"{key}:")
                      print(f"  Shape: {value.shape}")
                      print(f"  Range: [{value.min():.3f}, {value.max():.3f}]")
                      print(f"  Mean: {value.mean():.3f}")
                      print(f"  Std: {value.std():.3f}")
  
          # Extract and preprocess data
          events = sample_dict['events'].squeeze().numpy()
          image = sample_dict['image'].squeeze().numpy()
          depth = sample_dict['depth_image'].squeeze().numpy()
  
          # Save raw data for inspection
          print("\nSaving raw data samples...")
          np.save('raw_events.npy', events)
          np.save('raw_image.npy', image)
          np.save('raw_depth.npy', depth)
  
          # Create figure
          fig, axes = plt.subplots(2, 2, figsize=(15, 15))
  
          # Plot RGB image
          if len(image.shape) == 3:  # If image is RGB
              axes[0,0].imshow(normalize_image(image))
          else:  # If image is grayscale
              axes[0,0].imshow(normalize_image(image), cmap='gray')
          axes[0,0].set_title('RGB Image')
          axes[0,0].axis('off')
  
          # Plot depth with normalization
          depth_norm = normalize_image(depth)
          im = axes[0,1].imshow(depth_norm, cmap='plasma')
          plt.colorbar(im, ax=axes[0,1])
          axes[0,1].set_title(f'Depth (range: {depth.min():.2f} to {depth.max():.2f})')
          axes[0,1].axis('off')
  
          # Plot events sum with scaling
          event_sum = np.sum(events, axis=0)
          vmax = max(abs(event_sum.min()), abs(event_sum.max()))
          if vmax == 0:
              vmax = 1e-6
          im = axes[1,0].imshow(event_sum, cmap='RdBu_r', vmin=-vmax, vmax=vmax)
          plt.colorbar(im, ax=axes[1,0])
          axes[1,0].set_title(f'Events Sum (range: {event_sum.min():.2e} to {event_sum.max():.2e})')
          axes[1,0].axis('off')
  
          # Plot first event time bin with scaling
          event_bin = events[0]
          vmax = max(abs(event_bin.min()), abs(event_bin.max()))
          if vmax == 0:
              vmax = 1e-6
          im = axes[1,1].imshow(event_bin, cmap='RdBu_r', vmin=-vmax, vmax=vmax)
          plt.colorbar(im, ax=axes[1,1])
          axes[1,1].set_title(f'Events First Bin (range: {event_bin.min():.2e} to {event_bin.max():.2e})')
          axes[1,1].axis('off')
  
          plt.tight_layout()
          plt.savefig('sample_visualization.png', dpi=300, bbox_inches='tight')
          plt.close()
  
          print("\nVisualization saved as 'sample_visualization.png'")
          print("Raw data saved as 'raw_*.npy' files")
  
          # Print shape information
          print("\nData shapes:")
          print(f"Events shape: {events.shape}")
          print(f"Image shape: {image.shape}")
          print(f"Depth shape: {depth.shape}")
  
      except Exception as e:
          print(f"Error processing data: {str(e)}")
          import traceback
          traceback.print_exc()
  
  if __name__ == "__main__":
      check_data_structure()
      test_data()

``` 
The output from testcode is: 

Checking data structure...
✓ Found ./data/day1/train/depth/data
  Contains 1000 files
  First few files: ['depth_0000000000.npy', 'depth_0000000001.npy', 'depth_0000000002.npy', 'depth_0000000003.npy', 'depth_0000000004.npy']
✓ Found ./data/day1/train/events/voxels
  Contains 1000 files
  First few files: ['event_tensor_0000000000.npy', 'event_tensor_0000000001.npy', 'event_tensor_0000000002.npy', 'event_tensor_0000000003.npy', 'event_tensor_0000000004.npy']
✓ Found ./data/day1/train/rgb/davis
  Contains 1000 files
  First few files: ['frame_0000000000.png', 'frame_0000000001.png', 'frame_0000000002.png', 'frame_0000000003.png', 'frame_0000000004.png']

Building dataset...
entered concatenate_subfolders
Found 4 samples in ./data/day1/train
Dataset size: 11934

Testing first batch...

Data ranges:
events:
  Shape: torch.Size([1, 5, 260, 346])
  Range: [0.000, 0.000]
  Mean: 0.000
  Std: 0.000
image:
  Shape: torch.Size([1, 1, 260, 346])
  Range: [0.000, 1.000]
  Mean: 0.066
  Std: 0.097
depth_image:
  Shape: torch.Size([1, 1, 260, 346])
  Range: [nan, nan]
  Mean: nan
  Std: nan

Saving raw data samples...

Visualization saved as 'sample_visualization.png'
Raw data saved as 'raw_*.npy' files

Data shapes:
Events shape: (5, 260, 346)
Image shape: (260, 346)
Depth shape: (260, 346)

In this saved sample, the RGB and depth frames don't seem to match.

![sample_visualization](https://github.com/user-attachments/assets/e9b484e5-d949-489f-9988-9e6d12708009)
